src/gentropy/common/harmonise.py

```diff
@@ -0,0 +1,180 @@
+"""Variant harmonisation utilities."""
+
+import pyspark.sql.functions as f
+import pyspark.sql.types as t
+from pyspark.sql import DataFrame, SparkSession
+
+from gentropy.common.spark_helpers import neglog_pvalue_to_mantissa_and_exponent
+
+
+def harmonise_summary_stats(
+    spark: SparkSession,
+    raw_summary_stats_path: str,
+    tmp_variant_annotation_path: str,
+    chromosome: str,
+    colname_position: str,
+    colname_allele0: str,
+    colname_allele1: str,
+    colname_a1freq: str,
+    colname_info: str,
+    colname_beta: str,
+    colname_se: str,
+    colname_mlog10p: str,
+    colname_n: str,
+) -> DataFrame:
+    """Ingest and harmonise the summary stats.
+
+    1. Rename chromosome 23 to X.
+    2. Filter out low INFO rows.
+    3. Filter out low frequency rows.
+    4. Assign variant types.
+    5. Create variant ID for joining the variant annotation dataset.
+    6. Join with the Variant Annotation dataset.
+    7. Drop bad quality variants.
+
+    Args:
+        spark (SparkSession): Spark session object.
+        raw_summary_stats_path (str): Input raw summary stats path.
+        tmp_variant_annotation_path (str): Input variant annotation dataset path.
+        chromosome (str): Which chromosome to process.
+        colname_position (str): Column name for position.
+        colname_allele0 (str): Column name for allele0.
+        colname_allele1 (str): Column name for allele1.
+        colname_a1freq (str): Column name for allele1 frequency (optional).
+        colname_info (str): Column name for INFO, reflecting variant quality (optional).
+        colname_beta (str): Column name for beta.
+        colname_se (str): Column name for beta standard error.
+        colname_mlog10p (str): Column name for -log10(p).
+        colname_n (str): Column name for the number of samples.
+
+    Returns:
+        DataFrame: A harmonised summary stats dataframe.
+    """
+    # Read the precomputed variant annotation dataset.
+    va_df = (
+        spark
+        .read
+        .parquet(tmp_variant_annotation_path)
```

Comment:
What is the tmp_variant_annotation and what is the difference between it and other variant_annotation?

Reply:
Great question! `tmp_variant_annotation` is the Variant Annotation dataset which has been further prepared and processed by the per_chromosome module (previous PR in the chain) to speed up the joins in the harmonisation phase. It includes all variants in both the direct (A0/A1) and reverse (A1/A0) orientations, so that the direction of the variant can be easily determined on joining. I've also added the same description to the docstring.

Reply:
Yes, I realised it when I went to the next PR! It make sense, thank you for the explanation.